src/components/Layout/TableOfContents.jsx

```diff
@@ -1,165 +1,90 @@
-            <li>
-              <h6>{node.post.childMarkdownRemark.frontmatter.title}</h6>
-            </li>
-          </Link>
-        </LessonLIContainer>
-      )
-    })
-  }
-
-  buildChapterNodes(node) {
-    // If this is a Chapter (and not a subchapter)
-    if (this.currentLevel === 0) {
-      this.addChapterJSX(node);
-      // Else it's a SubChapter
-    } else {
-      this.addSubchapterJSX(node);
+import React from "react";
+import Link from "gatsby-link";
+import styled from "styled-components";
+
+const Links = ({ entries, index }) => (
+  <StyledLinkList>
+    {entries.map(({ entry }, key) => (
+      <EntryListItem key={`link${index}${key}`}>
+        <Link to={entry.childMarkdownRemark.fields.slug}>
+          <EntryTitle>{entry.childMarkdownRemark.frontmatter.title}</EntryTitle>
+        </Link>
+      </EntryListItem>
+    ))}
+  </StyledLinkList>
+);
+
+const Subchapters = ({ chapters, index }) => (
+  <StyledSubchapterList>
+    {chapters.map(({ entries, title }, key) => (
+      <li key={`subchapter${index}${key}`}>
+        <SubchapterTitle>{title}</SubchapterTitle>
+        <Links entries={entries} index={`subchapter${index}${key}`} />
+      </li>
+    ))}
+  </StyledSubchapterList>
+);
+
+const chapterMapper = ({ subchapters, entries, title }, index) => {
+  const AdditionalElements = () => {
+    if (subchapters) {
+      return <Subchapters chapters={subchapters} index={index} />;
     }
-  }
-
-
-  // Level matters because sub-chapter <li> and chapter <li>s are styled differently
-  //
-  // If the node is an Array, it holds the actual page nodes itself,
-  // 1. Add the node's value as either a chapter <li> or subchapter <li> - based on level
-  // 2. Build the lesson <li> links
-  //
-  // Else, its an object, so it must be a chapter or sub-chapter
-  // 1. build the Chapter <li>
-  // 2. Recursively restart on the child node
-  //
-  // For level, if we've hit an array, then we must be as deeply nested as possible.
-  // So the next node we hit will be a top level chapter node, so we're back to level 0.
-  //
-  buildNodes(nodes) {
-    function getNextNode(postNodes, node) {
-      const keys = Object.keys(nodes);
-      const nextIndex = keys.indexOf(node) +1;
-      return keys[nextIndex];
-    }
-
-    Object.keys(nodes).forEach(node => {
-      const nextNode = getNextNode(nodes, node);
-      if (Array.isArray(nodes[node])) {
-        // Add the Lowest Level Chapter Name (Title of Array):
-        this.buildChapterNodes(node);
-        this.buildLessonItemNodes(nodes[node]);
-        if (nextNode === undefined) {
-          this.currentLevel -= 1;
-        }
-      } else {
-        this.buildChapterNodes(node)
-        this.currentLevel += 1;
-        this.buildNodes(nodes[node])
-      }
-    });
-  }
-
-  render() {
-    const posts = this.props.posts.chapters;
-    this.nodeListItemsToRender = [];
-    this.buildNodes(posts);
-    return (
-      <TableOfContentsContainer>
-        <ul>
```

Comment:
The **order** entries in the menu for the TOC **matters**, I would say. So this should an **ordered list**. I also think that  using a "flat" list (chapter titles, subchapter titles, entry titles are all each just on `li` in the top level list) is a bad approach. I created sublists for the individual entries to have a more finegrained logic. This will be very helpful when someone wants to do more complex styling (some padding around the `EntryList`s for example).